torch/_dynamo/guards.py

```diff
@@ -414,9 +418,13 @@ def TENSOR_MATCH(self, guard: Guard):
             self.ID_MATCH(guard)
```

Comment:
noob question: what exactly does the `TENSOR_MATCH` method on this class do? It seems to generate some code, but when exactly does it called. (Also, what's a Source? There's no comments in that module.)

Reply:
The long form documentation is here - https://github.com/pytorch/torchdynamo/blob/main/documentation/GuardsOverviewPt1.md#making-guards - except with an `EQUALS_MATCH` example (logically analogous method, except for certain specialized constants)

Reply:
The doc says

> The source here is an enum indicating what kind of source the guard belongs to [Note: not to be confused with Source and the other types in source.py, as stored on VariableTracker, as discussed above]

Oh, knowing this is not source.py actually helps a lot. But... I still don't know what the enum means? So is global source, like, the guard arose due to a usage of a global variable?

Another thing that is a bit confusing is the class definition

```
@dataclasses.dataclass
class Guard:
    name: str
    source: GuardSource
    create_fn: Callable
```

so... naively, I expect a "guard" to be "an object representing a guard condition that specifies whether or not a given piece of code is valid to execute or not." So for any given compiled object I'd expect it to be associated with some number of Guard objects(?). But since Guard has a `create_fn`, I guess that's not what it does? So this what... creates guards? But there's also a guard builder?

What exactly a create fn is also confusing because naively, if I were creating a guard, I'd expect, you know, to have to pass in some information saying what I want the guard to actually do. And then when I create the guard, I use that info to drive the codegen. But `TENSOR_MATCH` doesn't actually take in any arguments besides guard? So I guess guard is the info? But aren't we building a guard (cuz this class is called GuardBuilder)? I guess GuardBuilder doesn't actually build Guards (so... rename it?) And now I see that there is some comments about how tensor matches get deferred, so I guess that helps a little? But it's all pretty murky.

Reply:
Discussed offline, perhaps the longform doc is overdue for a rewrite :)